module: add action_commands decorator

deathbybandaid opened the pull request:
As per my idea in Issue #1658, this is my attempt to integrate an `action_commands` command structure to Sopel.

There are still things to work out here, I'm sure. I copied elements I saw in `intents`, `commands`, and `nickname_commands`. I am very well aware the docstrings will definitely need work, as well as tests added.

I'm also not happy with the import line adjusted, and will likely fix that with `()` later.

This is more/less a draft PR, but my brief tests were very successful.

deathbybandaid commented:
Test file used:

````
# coding=utf-8

from __future__ import unicode_literals, absolute_import, division, print_function

import sopel.module


@sopel.module.commands('dbbtest')
def bot_command_dbbtest(bot, trigger):
    bot.say(str(trigger))
    bot.say(str(trigger.args))



@sopel.module.action_commands('rolls')
def bot_command_dbbtestb(bot, trigger):
    bot.say(str(trigger))
    bot.say(str(trigger.args))
````

Output:

````
<deathbybandaid> .dbbtest this out
<dbbbot> .dbbtest this out
<dbbbot> ['#deathbybandaid', '.dbbtest this out']


* deathbybandaid rolls dice
<dbbbot> rolls dice
<dbbbot> ['#deathbybandaid', 'rolls dice']

````

dgw commented:
Looking at this led me to the discovery that inspired #1683. Specifically, the implementation here made me wonder "what other intents are there?", which led to finding out that the intent spec no longer exists. This decorator is a pretty good idea, actually, especially _because_ we'll want to replace the internal references to "intents" on callables and triggers with something else that isn't obsolete. Could be as simple as `trigger.is_action` and a `callable.on_actions` Boolean (default `False`). Further discussion and ideas very much welcome in #1683!

deathbybandaid commented:
Forced Pushed to rebase with current master.

I could use some help writing the tests for `test_module.py`.

A thorough review would also be nice to clean/tidy it up. However, I feel fairly positive about what is accomplished here.

As discussed via IRC, this can be implemented, with the current `intents` code, with the intention (not a pun) of other PRs being opened later, such as:

* Re-evaluating how CTCP is handled
* Based on ^ how we want to handle something like `trigger.is_action`
* Deprecating `@intents`
* Removal of the ability to run `@commands` via a CTCP ACTION

dgw commented on sopel/module.py:
```diff
@@ -261,6 +261,30 @@ def add_attribute(function):
+    This decorator can be used multiple times to add multiple rules. The
+    resulting match object will have the command as the first group, rest of
+    the line, excluding leading whitespace, as the second group. Parameters 1
+    through 4, separated by whitespace, will be groups 3-6.
```
```suggestion
    through 4, separated by whitespace, will be groups 3-6.

```

dgw commented on sopel/module.py:
```diff
@@ -261,6 +261,30 @@ def add_attribute(function):
+    the line, excluding leading whitespace, as the second group. Parameters 1
+    through 4, separated by whitespace, will be groups 3-6.
+    Args:
+        command: A string, which can be a regular expression.
```
```suggestion
        command: A string, which can be a regular expression.

```

dgw commented on sopel/module.py:
```diff
@@ -261,6 +261,30 @@ def add_attribute(function):
+        command: A string, which can be a regular expression.
+    Returns:
+        A function with a new regular expression appended to the rule
+        attribute. If there is no rule attribute, it is added.
```
```suggestion
        attribute. If there is no rule attribute, it is added.

```

dgw commented on sopel/module.py:
```diff
@@ -261,6 +261,30 @@ def add_attribute(function):
+        attribute. If there is no rule attribute, it is added.
+    Example:
+        @action_commands("hello!"):
+            Would trigger on "/me hello!"
```
I think it would be clearer to give a "proper" action as example, e.g. `/me says hello`.

The explanation could then say it would trigger "when a user 'says hello' as an action", or something. We should avoid doing `"/me hello!"` or anything else that might imply "/me" is part of the message. (This is also why I suggested an amendment to the one-line description.)

dgw replied:
Meh. I'll leave this to be addressed when this file gets its turn for #1565.

deathbybandaid commented:
That was really odd with the whitespace disappearing from my push. Looked normal in my editor.

Added all the requested changes though.

dgw commented:
Hmm, that force-push undid all the changes. You should try rebasing/squashing again from 44cd95d.

deathbybandaid commented:
@dgw, something is screwy on my end, and everytime I squash, those whitespaces change.
I'll look at it more when I get the chance.

deathbybandaid commented:
I deleted the cloned directory, recloned, squashed, and it looks like the problem went away.

Not sure what was happening before, but it looks like everything is there.

dgw commented:
If I'm touching tests, I like @Exirel to look at it. 😁

dgw commented:
Had to rebase on `master` for that last tweak (getting logging & docstrings squared away in `bot.py`). I didn't _want_ to rebase, but the alternative was making a separate PR, and _that_ seemed silly.

dgw commented:
And rebased again due to conflicting changes in #1708. 😦

deathbybandaid commented on sopel/loader.py:
```diff
@@ -67,7 +68,7 @@ def clean_callable(func, config):
         func.rule = [compile_rule(nick, rule, alias_nicks) for rule in func.rule]
 
-    if hasattr(func, 'commands') or hasattr(func, 'nickname_commands'):
+    if hasattr(func, 'commands') or hasattr(func, 'nickname_commands') or hasattr(func, 'action_commands'):
```
I know it isn't that much shorter, but it might help line length here to do

`if all(hasattr(func, attr) for attr in ['commands', 'nickname_commands', 'action_commands']):`

dgw replied:
Using `all()` would break it.

Exirel replied:
Yes, `any` would be cool.
